// Backtesting Service - Runs historical strategy backtests using VectorBT and FinRL engines
// Part of BLOASIS Phase 1 infrastructure for strategy validation
//
// This service provides backtesting capabilities for trading strategies by:
// - Running historical simulations using VectorBT for vectorized operations
// - Supporting FinRL for reinforcement learning-based strategy evaluation
// - Computing performance metrics (Sharpe ratio, max drawdown, win rate)
// - Recording individual trades for detailed analysis
//
// Dependencies:
// - AI Analysis Service (Tier 3): Provides strategies to backtest
// - Market Data: Historical price data for simulation
//
// The backtesting workflow:
// 1. Receives strategy parameters (strategy_id, symbols, time range, capital)
// 2. Fetches historical market data for specified symbols and time range
// 3. Executes strategy logic using VectorBT/FinRL engines
// 4. Computes performance metrics and trade history
// 5. Returns comprehensive backtest results

syntax = "proto3";

package bloasis.backtesting;

import "google/api/annotations.proto";
import "common.proto";

// BacktestingService provides historical strategy backtesting capabilities
// Uses VectorBT for fast vectorized backtesting and FinRL for RL-based evaluation
service BacktestingService {
  // BacktestStrategy runs a historical backtest for a given strategy
  // Uses VectorBT and FinRL engines for efficient simulation
  //
  // The backtest process:
  // 1. Validates input parameters and strategy existence
  // 2. Fetches historical data for symbols within time_range
  // 3. Executes strategy with initial_capital
  // 4. Computes comprehensive performance metrics
  // 5. Records all trades for detailed analysis
  //
  // Note: This is a computationally intensive operation
  // Typical execution time: 5-30 seconds depending on time range and symbol count
  rpc BacktestStrategy(BacktestStrategyRequest) returns (BacktestStrategyResponse) {
    option (google.api.http) = {
      post: "/v1/backtesting/backtest-strategy"
      body: "*"
    };
  }
}

// BacktestStrategyRequest - Request parameters for running a strategy backtest
message BacktestStrategyRequest {
  // Unique identifier of the strategy to backtest
  // Typically generated by AI Analysis Service (Tier 3)
  // Format: "strategy_{user_id}_{timestamp}"
  string strategy_id = 1;

  // User identifier who owns this strategy
  // Used for tracking and audit purposes
  string user_id = 2;

  // Symbols to include in the backtest simulation
  // Stock ticker symbols (e.g., ["AAPL", "GOOGL", "MSFT"])
  // Recommendation: 1-50 symbols per backtest for optimal performance
  repeated string symbols = 3;

  // Time range for historical simulation
  // Uses bloasis.common.TimeRange for ISO 8601 date format
  // Example: start_date="2024-01-01T00:00:00Z", end_date="2024-12-31T23:59:59Z"
  // Note: Longer ranges require more computation time
  bloasis.common.TimeRange time_range = 4;

  // Initial capital for the backtest simulation
  // Uses bloasis.common.Money for Decimal precision (critical for accurate backtesting)
  // Example: amount="100000.00", currency="USD"
  bloasis.common.Money initial_capital = 5;
}

// BacktestStrategyResponse - Comprehensive backtest results
message BacktestStrategyResponse {
  // Unique identifier for this backtest run
  // Format: "backtest_{strategy_id}_{timestamp}"
  string backtest_id = 1;

  // Strategy identifier that was backtested
  // Matches the strategy_id from the request
  string strategy_id = 2;

  // Performance metrics computed from the backtest
  // Includes returns, risk metrics, and trade statistics
  BacktestMetrics metrics = 3;

  // Individual trades executed during the backtest
  // Ordered chronologically (oldest first)
  // Includes all BUY and SELL transactions
  repeated BacktestTrade trades = 4;

  // ISO 8601 timestamp when this backtest was completed
  // Example: "2025-01-26T14:30:00Z"
  string timestamp = 5;
}

// BacktestMetrics - Performance metrics from a backtest simulation
// All percentage values are expressed as decimals (e.g., 0.15 = 15%)
message BacktestMetrics {
  // Total return percentage over the backtest period
  // Calculated as: (final_value - initial_capital) / initial_capital
  // Example: 0.25 represents 25% total return
  double total_return = 1;

  // Sharpe ratio - risk-adjusted return metric
  // Calculated as: (return - risk_free_rate) / standard_deviation
  // Higher values indicate better risk-adjusted performance
  // Typical interpretation: < 1 = subpar, 1-2 = good, > 2 = excellent
  double sharpe_ratio = 2;

  // Maximum drawdown percentage - largest peak-to-trough decline
  // Measures the worst historical loss from a peak
  // Example: 0.20 represents 20% max drawdown
  // Lower values indicate better downside protection
  double max_drawdown = 3;

  // Win rate percentage - proportion of profitable trades
  // Calculated as: profitable_trades / total_trades
  // Example: 0.55 represents 55% win rate
  double win_rate = 4;

  // Total number of trades executed during the backtest
  // Includes both BUY and SELL transactions
  int32 total_trades = 5;

  // Final portfolio value at the end of the backtest period
  // Uses bloasis.common.Money for Decimal precision
  // Compare with initial_capital to verify total_return calculation
  bloasis.common.Money final_value = 6;
}

// BacktestTrade - Individual trade record from a backtest simulation
message BacktestTrade {
  // Stock ticker symbol for this trade (e.g., "AAPL", "GOOGL")
  string symbol = 1;

  // Trade action:
  // - "BUY": Open or add to long position
  // - "SELL": Close or reduce long position
  string action = 2;

  // Number of shares traded
  // Always positive regardless of action type
  int32 quantity = 3;

  // Execution price per share
  // Uses bloasis.common.Money for Decimal precision (critical for accurate P&L)
  // Represents the simulated fill price from historical data
  bloasis.common.Money price = 4;

  // ISO 8601 timestamp when this trade was executed
  // Example: "2024-06-15T14:30:00Z"
  string timestamp = 5;
}
